libkernelflinger/tpm2_security.c

```diff
@@ -107,6 +107,60 @@ static const attribute_matrix_t config_table[MAX_NV_NUMBER] = {
 	}
 };
 
+
+static EFI_STATUS tpm2_get_capability(
+		IN      TPM_CAP                   Capability,
+		IN      UINT32                    Property,
+		IN      UINT32                    PropertyCount,
+		OUT     TPMI_YES_NO               * MoreData,
+		OUT     TPMS_CAPABILITY_DATA      * CapabilityData
+		)
+{
+	EFI_STATUS ret;
+	UINT32 i;
+
+	ret = Tpm2GetCapability(Capability, Property, PropertyCount, MoreData, CapabilityData);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Call Tpm2GetCapability failed");
+		return ret;
+	}
+	debug(L"TPM2 capability: 0x%08x, data.tpmProperties.count: %d, more data: %d",
+			SwapBytes32(CapabilityData->capability), SwapBytes32(CapabilityData->data.tpmProperties.count), *MoreData);
+	for (i = 0; i < SwapBytes32(CapabilityData->data.tpmProperties.count); i++) {
+		debug(L"prop %d: property: 0x%08x, value: 0x%08x", i,
+				SwapBytes32(CapabilityData->data.tpmProperties.tpmProperty[i].property),
+				SwapBytes32(CapabilityData->data.tpmProperties.tpmProperty[i].value));
+	}
+
+	return ret;
+}
+
+static EFI_STATUS tpm2_get_cap_permanent(TPMA_PERMANENT *per)
+{
+	EFI_STATUS ret;
+	TPMI_YES_NO more_data;
+	TPMS_CAPABILITY_DATA cap_data;
+	UINT32 value;
+
+	ret = tpm2_get_capability(TPM_CAP_TPM_PROPERTIES, TPM_PT_PERMANENT, 1, &more_data, &cap_data);
+	if (EFI_ERROR(ret)) {
+		efi_perror(ret, L"Get TPM cap permanent failed");
+		return ret;
```

Comment:
Same than before: an intermediate variable for `cap_data.data.tpmProperties` should simplify the code